packages/api/src/command/medical/cq-directory/search-cq-directory.ts

```diff
@@ -0,0 +1,106 @@
+import convert from "convert-units";
+import { Sequelize } from "sequelize";
+import { Coordinates, getCoordinates } from "../../../external/location/address";
+import { CQDirectoryEntryModel } from "../../../models/medical/cq-directory";
+import { getPatientOrFail } from "../patient/get-patient";
+
+export const DEFAULT_RADIUS_IN_MILES = 50;
+
+export type CQOrgBasicDetails = {
+  name: string | undefined;
+  oid: string;
+  lon: number | undefined;
+  lat: number | undefined;
+  urlXCPD: string;
+  urlDQ: string | undefined;
+  urlDR: string | undefined;
+};
+
+/**
+ * Searches the Carequality Directory for organizations within a specified radius of a patient's addresses.
+ * @param cxId The ID of the customer organization.
+ * @param patientId The ID of the patient.
+ * @param radiusInMiles Optional, the radius in miles within which to search for organizations. Defaults to 50 miles.
+ *
+ * @returns Returns the details of organizations within the specified radius of the patient's address.
+ */
+export const searchNearbyCQOrganizations = async ({
+  cxId,
+  patientId,
+  radiusInMiles = DEFAULT_RADIUS_IN_MILES,
+}: {
+  cxId: string;
+  patientId: string;
+  radiusInMiles?: number;
+}) => {
+  const radiusInMeters = convert(radiusInMiles).from("mi").to("m");
+
+  const patient = await getPatientOrFail({ id: patientId, cxId });
+  const coordinates = await getCoordinates(patient.data.address);
+
+  const orgs = await searchCQDirectoriesByRadius({
+    coordinates,
+    radiusInMeters,
+  });
+
+  return pickBasicOrganizationAttributes(orgs);
+};
+
+/**
+ * Searches the Carequality Directory for organizations within a specified radius around geographic coordinates.
+ *
+ * @param coordinates The latitude and longitude around which to search for organizations.
+ * @param radiusInMeters The radius in meters within which to search for organizations.
+ * @returns Returns organizations within the specified radius of the patient's address.
+ */
+export const searchCQDirectoriesByRadius = async ({
+  coordinates,
+  radiusInMeters,
+}: {
+  coordinates: Coordinates[];
+  radiusInMeters: number;
+}): Promise<CQDirectoryEntryModel[]> => {
+  const orgs = [];
+
+  // TODO: MAKE SURE LAT LON WITHIN RANGE
+  for (const coord of coordinates) {
+    const orgsForAddress = await CQDirectoryEntryModel.findAll({
+      attributes: {
+        include: [
+          [
+            Sequelize.literal(
+              `ROUND(earth_distance(ll_to_earth(${coord.lat}, ${coord.lon}), ll_to_earth(lat, lon))::NUMERIC, 2)`
+            ),
+            "distance",
+          ],
+        ],
+      },
+      where:
+        Sequelize.literal(`earth_box(ll_to_earth (${coord.lat}, ${coord.lon}), ${radiusInMeters}) @> ll_to_earth (lat, lon)
+        AND earth_distance(ll_to_earth (${coord.lat}, ${coord.lon}), ll_to_earth (lat, lon)) < ${radiusInMeters}`),
```

Comment:
This might be something we can do some improvement as well. I'm assuming this is doing a full-table scan, right?

We probably can't do much about it since the filter criteria is applying a function on the columns with dynamic values - still worth checking how `earth_box` and `earth_distance` work WRT indexes, but maybe we can store `ll_to_earth(lat, lon)` as a precomputed column. It might not help much, but just something that crossed my mind as well.

IIRC, we don't have indexes on `lat`, `long`, but we don't do any query on those without applying a function to the values, right? If so, unless Postgres' geolocation functions can leverage it, not worth adding those.

Reply:
hmm that's a good point.. storing that instead of, or in addition to, the coordinates could save some time when querying for each patient's nearby facilities

Reply:
I was looking to see if it's worth the time (I'd recommend you do too), and TIL Posgres has a type of index that supports complex types, it's called GiST: https://blog.devgenius.io/generalized-search-tree-gist-indexes-in-postgresql-ba77e1bf26a1

Found about it here (good reading): https://dba.stackexchange.com/questions/314984/optimizing-geolocation-query-in-postgres-with-earth-box-and-earth-distance

Found this too: https://stackoverflow.com/questions/18059740/how-to-create-a-postgresql-index-that-includes-latitude-longitude-using-gist

I'd look into these to better understand how it works, best practices, and caveats, while building this type of solution.